src/ModelIO/MDLAnimatedValueTypes.cs

```diff
@@ -0,0 +1,877 @@
+using XamCore.Foundation;
+using XamCore.ObjCRuntime;
+
+using Vector2 = global::OpenTK.Vector2;
+using Vector2d = global::OpenTK.Vector2d;
+using Vector3 = global::OpenTK.NVector3;
+using Vector3d = global::OpenTK.NVector3d;
+using Vector4 = global::OpenTK.Vector4;
+using Vector4d = global::OpenTK.Vector4d;
+using Matrix4 = global::OpenTK.NMatrix4;
+using Matrix4d = global::OpenTK.NMatrix4d;
+using Quaternion = global::OpenTK.Quaternion;
+using Quaterniond = global::OpenTK.Quaterniond;
+
+// Because of the lack of documentation when at the time of binding ModelIO for iOS 11 types
+// we are implementing something similar to what swift is doing here
+// https://github.com/apple/swift/blob/cbdf0ff1e7bfbd192c33d64c9c7d31fbb11f712c/stdlib/public/SDK/ModelIO/ModelIO.swift
+
+namespace XamCore.ModelIO {
+	public partial class MDLAnimatedValue {
+
+		public double [] KeyTimes {
+			get {
+				var wkt = WeakKeyTimes;
+				if (wkt == null)
+					return null;
+
+				var count = wkt.Length;
+				var ret = new double [count];
+
+				for (int i = 0; i < count; i++)
+					ret [i] = wkt [i].DoubleValue;
+
+				return ret;
+			}
+		}
+
+		// Follow suit with swift's but making this
+		// a method instead of a property for three reasons:
+		// 1. Debugger performance.
+		// 2. _GetTimes return value is ignored and could turn out useful at some point.
+		// 3. Lack of documentation at the moment of binding this.
+		// [1]: https://github.com/apple/swift/blob/cbdf0ff1e7bfbd192c33d64c9c7d31fbb11f712c/stdlib/public/SDK/ModelIO/ModelIO.swift#L50
+		public virtual double [] GetTimes ()
+		{
+			var count = TimeSampleCount;
+			var timesArr = new double [(int) count];
+			unsafe {
+				fixed (double* ptr = timesArr)
+					_GetTimes ((IntPtr) ptr, count);
+			}
+			return timesArr;
+		}
+	}
+
+	public partial class MDLAnimatedScalarArray {
+
+		public virtual void SetValues (float [] array, double time)
+		{
+			if (array == null)
+				throw new ArgumentNullException (nameof (array));
+
+			unsafe {
+				fixed (float* ptr = array)
+					_SetFloatArray ((IntPtr) ptr, (nuint) array.Length, time);
+			}
+		}
+
+		public virtual void SetValues (double [] array, double time)
+		{
+			if (array == null)
+				throw new ArgumentNullException (nameof (array));
+
+			unsafe {
+				fixed (double* ptr = array)
+					_SetDoubleArray ((IntPtr) ptr, (nuint) array.Length, time);
+			}
+		}
+
+		public virtual float [] GetFloatValues (double time)
+		{
+			var count = ElementCount;
+			var timesArr = new float [(int) count];
+
+			unsafe {
+				fixed (float* ptr = timesArr)
+					_GetFloatArray ((IntPtr) ptr, count);
+			}
+			return timesArr;
+		}
+
+		public virtual double [] GetDoubleValues (double time)
+		{
+			var count = ElementCount;
+			var timesArr = new double [(int) count];
+
+			unsafe {
+				fixed (double* ptr = timesArr)
+					_GetDoubleArray ((IntPtr) ptr, count);
+			}
+			return timesArr;
+		}
+
+		public virtual void Reset (float [] values, double [] times)
```

Comment:
In Reset methods in later classes you're comparing the lengths of the values and times arrays as well as null checking both.  Should that be happening here as well?  There's quite a few without the check, then quite a few more with it.

Reply:
No, this is expected, in the non array container versions (like `MDLAnimatedVector3`) both arrays must match but in the array versions (like `MDLAnimatedVector3Array`) arrays can have different length.